tests/mcts/test_mcts.py

```diff
@@ -0,0 +1,113 @@
+    problem = QEDOptimizationProblem(config)
+    game = MCTS(problem)
+    return game
+
+
+def test_reward(game):
+    root = game._get_root()
+    assert game.compute_reward(root) == 0.0
+
+    game._expand(root)
+
+    assert game.compute_reward(root) == 0.0
+    assert game.compute_reward(root.children[-1]) == 0.3597849378839701
+    assert game.compute_reward(root) == 0.0
+
+
+def test_ucb_score(game):
+    root = game._get_root()
+    game._expand(root)
+    child = root.children[0]
+
+    root.update(2.0)
+    assert game._ucb_score(root, child) == math.inf
+    assert root.visit_count == 1
+    assert root.value == pytest.approx(2.)
+
+    root.update(4.0)
+    assert game._ucb_score(root, child) == math.inf
+    assert root.visit_count == 2
+    assert root.value == pytest.approx(3.)
+
+    game._backpropagate([root, child], 3.0)
+    assert game._ucb_score(root, child) == pytest.approx(5.09629414793641)
+
+
+def test_get_successors(game):
+    root = game._get_root()
+    successors = game._expand(root)
+    assert len(successors) == 9
+    assert successors[-1].state.is_terminal
+    assert not successors[0].state.is_terminal
+
+# @pytest.mark.parametrize('qed_case', ['mcts', 'nx_mcts', 'az'], indirect=True)
+# def test_tree_policy(qed_case):
+#     assert len(list(qed_case.tree_policy())) == 1
+#
+# def test_tree_policy(qed_root):
+#     assert len(list(qed_root.tree_policy())) == 1
+#     qed_case.update(1.)
+#     qed_case.expand()
+#     assert len(list(qed_case.tree_policy())) == 2
+#
+#     qed_root.update(1.)
+#     qed_root.expand()
+#     assert len(list(qed_root.tree_policy())) == 2
+#
+#     for child in qed_root.successors:
+#     for child in qed_case.successors:
+#         child.update(1.)
+#         child.expand()
+#
+#     assert len(list(qed_root.tree_policy())) == 3
+#     assert len(list(qed_case.tree_policy())) == 3
+#
+# @pytest.mark.parametrize('qed_case', ['mcts', 'nx_mcts', 'az'], indirect=True)
+# def test_evaluate(qed_case):
+#     assert qed_case.evaluate() > 0.
+#
+# def test_evaluate(qed_root):
+#     random.seed(42)
+#     assert qed_root.evaluate() == pytest.approx(0.4521166661434767)
+#
+#
+# def test_mcts_step(qed_root):
+# @pytest.mark.parametrize('qed_case', ['mcts', 'nx_mcts', 'az'], indirect=True)
+# def test_mcts_step(qed_case):
+#     random.seed(42)
+#     for _ in range(10):
+#         qed_root.mcts_step()
+#         qed_case.mcts_step()
+#
+#     assert qed_root.visits == 10
+#     assert qed_root.value == pytest.approx(0.37061839971331995)
+#     assert qed_case.visits == 10
+#     assert qed_case.value > 0.1
+#
+#
+# def test_run_mcts(qed_root):
+# @pytest.mark.parametrize('qed_case', ['mcts', 'nx_mcts', 'az'], indirect=True)
+# def test_run_mcts(qed_case):
+#     random.seed(42)
+#     history = list(qed_root.run_mcts(10, explore=False))
+#     assert history[-1].reward > 0.4
+#     history = list(qed_case.run_mcts(5, explore=False))
+#     assert history[-1].reward > 0.3
```

Comment:
Here, for instance, was where it was nice to have an inherited implementation between mcts and alphazero